### ISSUE TITLE
Constructing a ParameterDict raises a warning

### ISSUE THREAD

**Issue by fritzo** [pytorch/pytorch#46983]
## 🐛 Bug

Constructing a `ParameterDict` raises a `UserWarning`.

## To Reproduce

Install PyTorch 1.7
```py
>>> import torch.nn as nn
>>> nn.ParameterDict()
/Users/fobermey/opt/miniconda3/envs/pyro/lib/python3.7/site-packages/torch/nn/modules/container.py:550: UserWarning: Setting attributes on ParameterDict is not supported.
  warnings.warn("Setting attributes on ParameterDict is not supported.")
ParameterDict()
```

## Expected behavior

PyTorch should not raise warnings in normal usage. Tests should verify this.

## Environment

```
PyTorch version: 1.7.0
Is debug build: True
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Mac OSX 10.15.7 (x86_64)
GCC version: Could not collect
Clang version: 12.0.0 (clang-1200.0.32.2)
CMake version: version 3.18.2

Python version: 3.7 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.1
[pip3] numpyro==0.3.0
[pip3] torch==1.7.0
[pip3] torchfile==0.1.0
[pip3] torchvision==0.8.1
[conda] numpy                     1.19.0                   pypi_0    pypi
[conda] numpyro                   0.3.0                    pypi_0    pypi
[conda] torch                     1.7.0                    pypi_0    pypi
[conda] torchfile                 0.1.0                    pypi_0    pypi
[conda] torchvision               0.8.1                    pypi_0    pypi
```

cc @ezyang @gchanan @zou3519 @bdhirsh @heitorschueroff @albanD @mruberry

**Comment by gchanan** [comment]
Also happens for `ParameterList`

**Comment by gchanan** [comment]
it's coming from the top-level Module.

**Comment by gchanan** [comment]
CC @albanD can you take a look?

**Comment by albanD** [comment]
Indeed since ParameterList inherits from the Module class and the Module class sets attributes (like `.training`) we have this warning. Not sure how this was not caught before :/ 
I think we should only enable the warning after the `super(ParameterList, self).__init__()` has been called.
I guess this is one more quirk of this object being a nn.Module but not really.

**Comment by OmranKaddah** [comment]
@albanD I placed the warning after `super(ParameterList, self).__init__() `has been called
I am still getting the warning for using ParameterList. 
Should I just ignore it? It does not affect anything else?

**Comment by albanD** [comment]
If the warning happens after the initialization, there might be something wrong with your code.
In particular, everything that is set on these Modules that are not Parameters cannot be retrieved. So:
- If you set something without ever reading it. You can ignore the warning. And you can avoid setting that attribute altogether to avoid the warning
- If you set something that you read later. You have an issue as the attribute will always be reported as not existing!

**Comment by OmranKaddah** [comment]
I edited this way.
```
    def __setattr__(self, key: Any, value: Any) -> None:
        super(ParameterList, self).__setattr__(key, value)
        if not isinstance(value, torch.nn.Parameter):
            warnings.warn("Setting attributes on ParameterList is not supported.")

```
I do not understand you well.  First of all, the model that I am using did not show these warnings in the previous version with Cuda 10.2. I am not adding anything else, for sure, to `ParameterList` other than an initialized `Parameter`. The error repeatedly appears as each `Parameter` in the list is read.

The code I am using:
```
def uniform(size, tensor):
    bound = 1.0 / math.sqrt(size)
    if tensor is not None:
        tensor.data.uniform_(-bound, bound)
class MyModel(torch.nn.Module):
    def __init__(self, out_channels):
        super(MyModel, self).__init__()
        self.weights = torch.nn.ParameterList()
        for _ in range(3):
            weight = torch.nn.Parameter(torch.Tensor(out_channels, out_channels))
            uniform(out_channels,weight)
            self.weights.append(weight)

    def forward(self, x):
        for i in range(3):
            x = torch.matmul(self.weights[i], torch.sigmoid(x))
        return x

model = MyModel(2)
in_tensor = torch.tensor([[1.0, 2.0],[3.0, 4.0]])
print(model(in_tensor))

```
The example here is also giving the error:
https://pytorch.org/docs/stable/generated/torch.nn.ParameterList.html

**Comment by albanD** [comment]
> I edited it this way.

I meant a different update: only add this warning after `super(ParameterList, self).__init__()` (notice that this is the `__init__` function, not the `__setattr__` one) has been called.

**Comment by OmranKaddah** [comment]
My bad. It is fine now

**Comment by CedricPicron** [comment]
I believe the current fix does not work in all cases. The warning for example still incorrectly triggers when `model.train()` is called (after initialization) on a `model` containing ParametersLists/Dicts, as their `.training` attributes will be set.

**Comment by albanD** [comment]
That a very good point. Thanks for catching that.
We should update `train`, `eval` and `__setstate__` to disable the warning before calling into the original functions.

Could you send a PR for that by any chance?

**Comment by CedricPicron** [comment]
> We should update train, eval and __setstate__ to disable the warning before calling into the original functions.

Alternatively, we could add  the `not hasattr(self, key)` check to
https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/container.py#L434
which should avoid triggering the warning when general `Module` attributes such as `.training` are set. @albanD Tell me whether you like this solution.

> Could you send a PR for that by any chance?

Sure, I'll do it.

**Comment by albanD** [comment]
That would not work to avoid the warnings in `__setstate__` though as it creates brand new attributes.
But looking through the whole Module code, only these 3 functions actually set states (beyond the `__init__`).
So I would say that overwriting these functions to disable the warning should work for now.

> Sure, I'll do it.

Awesome! Just add me as a reviewer!

**Comment by CedricPicron** [comment]
> That would not work to avoid the warnings in __setstate__ though as it creates brand new attributes.

Yes, I saw that. However, I couldn't come up with a scenario were these attributes would be missing as these are set when calling `super(ParameterList, self).__init__()`. As far as I understand, these `__setstate__` exceptions should only trigger if you create a Module from an old checkpoint without first initializing it through `super().__init__()`.

**Comment by albanD** [comment]
Ho the whole point of this function is to make sure we can load checkpoints that come from older versions of pytorch. And so for which these attributes don't actually exist.

**Comment by albanD** [comment]
Yes the whole point of this code is to make sure that even when people save the whole module (and not just the state dict as they should), they can reload it.
And in this case, because the old module didn't had these attributes, this function will create them and would raise the warning.

**Comment by sbocconi** [comment]
Calling `model.eval()` or `model.train()` still causes `Setting attributes on ParameterDict is not supported.` with torch `1.7.1.post2` on Mac 10.15.7.

Warning is in container.py, `__setattr__` function on line 552.

**Comment by albanD** [comment]
Hi,

Running this does not warn for me on master and on 1.8
```python
import torch
print(torch.__version__)
from torch import nn

class Mod(nn.Module):
    def __init__(self):
        super().__init__()
        self.foo = nn.ParameterList([nn.Parameter(torch.rand(10))])

mod = Mod()
mod.eval()
mod.train()
```

Do you have a code sample that repro what you observe?

**Comment by sbocconi** [comment]
It does for me, so probably I need to update to 1.8? I am on torch 1.7.1.post2 as reported above.

**Comment by albanD** [comment]
Yes not all the fix went into 1.7.1 unfortunately.
Note that this warning is harmless, the beyond the spam, you can safely ignore it.
To get rid of it, you can try out the nightly build to get the fix. Or wait a few hours and get the brand new 1.8 release.

**Comment by sbocconi** [comment]
Thanks, I am ignoring it now with:

```
import warnings
warnings.filterwarnings("ignore", message="Setting attributes on ParameterDict is not supported.")
```

I will wait for the new release!!
